src/modules/musig/musig.md

```diff
@@ -1,198 +1,59 @@
+2. The `secp256k1_musig_secnonce` structure is never copied or serialized.
+   See also the comment on `secp256k1_musig_secnonce` in `include/secp256k1_musig.h`.
+3. Opaque data structures are never written to or read from directly.
+   Instead, only the provided accessor functions are used.
+4. If adaptor signatures are used, all partial signatures are verified.
 
-where `P_i` is the public key of the `i`th signer and `µ_i` is a so-called
-_MuSig coefficient_ computed according to the following equation
+# Key Aggregation and (Taproot) Tweaking
 
-    L = H(P_1 || P_2 || ... || P_n)
-    µ_i = H(L || i)
+Given a set of public keys, the aggregate public key is computed with `secp256k1_musig_pubkey_agg`.
+A (Taproot) tweak can be added to the resulting public key with `secp256k1_xonly_pubkey_tweak_add`.
 
-where H is a hash function modelled as a random oracle.
+# Signing
 
-To produce a multisignature `(s, R)` on a message `m` using verification key
-`P`, signers act as follows:
+This is covered by `src/modules/musig/example.c`.
+Essentially, the protocol proceeds in the following steps:
 
-1. Each computes a nonce, or ephemeral keypair, `(k_i, R_i)`. Every signer
-   communicates `H(R_i)` to every participant (both signers and auditors).
-2. Upon receipt of every `H(R_i)`, each signer communicates `R_i` to every
-   participant. The recipients check that each `R_i` is consistent with the
-   previously-communicated hash.
-3. Each signer computes a combined nonce
-    `R = sum_i R_i`
-   and shared challenge
-    `e = H(R || P || m)`
-   and partial signature
-    `s_i = k_i + µ_i*x_i*e`
-   where `x_i` is the secret key corresponding to `P_i`.
+1. Generate a keypair with `secp256k1_keypair_create` and obtain the xonly public key with `secp256k1_keypair_xonly_pub`.
+2. Call `secp256k1_musig_pubkey_agg` with the xonly pubkeys of all participants.
+3. Optionally add a (Taproot) tweak with `secp256k1_musig_pubkey_tweak_add`.
+4. Generate a pair of secret and public nonce with `secp256k1_musig_nonce_gen` and send the public nonce to the other signers.
+5. Someone (not necessarily the signer) aggregates the public nonce with `secp256k1_musig_nonce_agg` and sends it to the signers.
+6. Process the aggregate nonce with `secp256k1_musig_nonce_process`.
+7. Create a partial signature with `secp256k1_musig_partial_sign`.
+8. Verify the partial signatures (optional in some scenarios) with `secp256k1_musig_partial_sig_verify`.
+9. Someone (not necessarily the signer) obtains all partial signatures and aggregates them into the final Schnorr signature using `secp256k1_musig_partial_sig_agg`.
 
-The complete signature is then the `(s, R)` where `s = sum_i s_i` and `R = sum_i R_i`.
+The aggregate signature can be verified with `secp256k1_schnorrsig_verify`.
+Note that steps 1 to 3 can happen after generating the nonce in step 4, but this disables some of the defense-in-depth measures.
```

Comment:
Is it also worth mentioning that steps 1-6 can happen during a "preprocessing step" (i.e. without knowing the message)?

Reply:
Good idea, I rephrased that sentence.